File: frontend/src/types/project.ts
import type { TagResponse } from "./tag";

export interface Project {
  id: number;
  owner_id: number;
  name: string;
  description: string | null;
  cover_path: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
  is_favorite?: boolean;
  is_shared?: boolean;
  my_role?: "owner" | "editor" | "viewer" | "commenter" | "admin";
  owner_name?: string | null;
  owner_email?: string | null;
  tags: TagResponse[];
  video_count: number;
  recent_project: string | null;
  duration: string | null;
  size: string | null;
}

export interface ProjectFavoriteResponse {
  project_id: number;
  is_favorite: boolean;
}

export interface ProjectMember {
  id: number;
  project_id: number;
  user_id?: number | null;
  email: string;
  role: "viewer" | "commenter" | "editor" | "admin";
  status: "pending" | "accepted" | "declined" | "revoked";
  created_at: string;
  full_name?: string | null;
}

export interface ProjectMemberAddRequest {
  email: string;
  role: "viewer" | "commenter" | "editor" | "admin";
}

export interface ProjectMemberUpdateRequest {
  role: "viewer" | "commenter" | "editor" | "admin";
}

export interface ProjectCreateRequest {
  name: string;
  description?: string | null;
  cover_path?: string | null;
  tag_ids?: number[] | null;
}

export interface ProjectUpdateRequest {
  name?: string | null;
  description?: string | null;
  cover_path?: string | null;
  status?: string | null;
}

export interface ProjectFolder {
  id: number;
  project_id: number;
  parent_id?: number | null;
  name: string;
  created_at: string;
  updated_at: string;
}

export interface FolderCreateRequest {
  name: string;
  parent_id?: number | null;
}

export interface FolderUpdateRequest {
  name?: string;
  parent_id?: number | null;
}

export type AssetCategory =
  | "all"
  | "video"
  | "audio"
  | "subtitle"
  | "transcript"
  | "document"
  | "speaker_voice";

export interface ProjectAssetItem {
  id: string;
  name: string;
  category: "video" | "audio" | "subtitle" | "transcript" | "document" | "speaker_voice" | string;
  format?: string;
  size_bytes?: number;
  size_display?: string;
  storage_location?: string;
  storage_type?: string;
  path_or_key: string;
  download_url?: string | null;
  preview_url?: string | null;
  created_at?: string | null;
  folder_id?: number | null;
  folder_name?: string | null;
  video_id?: number | string | null;
  video_title?: string | null;
}

export interface ProjectAssetsResponse {
  project_id: number;
  total_files: number;
  total_size_bytes: number;
  category_counts: Record<string, number>;
  assets: ProjectAssetItem[];
}

File: backend/app/schemas/project.py
from datetime import datetime
from pydantic import BaseModel, Field

from app.schemas.tag import TagResponse


# =========================================================
# Create Project
# =========================================================

class ProjectCreateRequest(BaseModel):
    name: str = Field(
        min_length=1,
        max_length=255,
    )
    description: str | None = Field(
        default=None,
    )
    cover_path: str | None = Field(
        default=None,
    )
    tag_ids: list[int] | None = Field(
        default=None,
    )


# =========================================================
# Update Project
# =========================================================

class ProjectUpdateRequest(BaseModel):
    name: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
    )
    description: str | None = Field(
        default=None,
    )
    cover_path: str | None = Field(
        default=None,
    )
    status: str | None = Field(
        default=None,
        max_length=50,
    )


# =========================================================
# Project Response
# =========================================================

class ProjectResponse(BaseModel):
    id: int
    owner_id: int
    name: str
    description: str | None
    cover_path: str | None
    status: str
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None
    is_favorite: bool = False
    is_shared: bool = False
    my_role: str = "owner"
    owner_name: str | None = None
    owner_email: str | None = None
    tags: list[TagResponse] = []
    video_count: int = 0
    recent_project: str | None = None
    duration: str | None = None
    size: str | None = None


# =========================================================
# Project Favorite Response
# =========================================================

class ProjectFavoriteResponse(BaseModel):
    project_id: int
    is_favorite: bool



